### PR TITLE
editor: Improve Evy web editor to use smarter indentation

### PR DESCRIPTION
**Pull request by juliaogris**
Use smarter indentation in Evy web editor.

Whenever a new line is inserted, the editor will automatically:

- preserves indent of current line by default (we already had this).
- increases the indent after block opening keywords
  `func` `on` `if` `else` `while` `for` (new).
- updates the current line indent for `end` and `else` keywords:
  decreasing the indent if needed (new).

There are many edge cases. This is a best effort improvement. Current
line updating only happens if no selection is active - selection start
and end index are equal.

Fixes: https://github.com/evylang/todo/issues/115

---

The main challenge of this PR is to a bunch of tedious manual testing of the editor.

I will try and write a longer Evy program in the web editor before merging.

### CODE REVIEW

File: frontend/module/editor.js [hunk 1]
```diff
@@ -343,25 +346,68 @@ const preserveIndent = () => (textareaProps, event) => {
 
   const lines = value.split("\n")
   const currentLine = lines[currentLineNumber]
-  const matches = /^\s+/.exec(currentLine)
-
-  if (!matches) {
+  const matches = /^(\s+)?((func|on|if|while|for)\b)?(end|else\b)?/.exec(currentLine)
```
Comment (camh-) [inline frontend/module/editor.js:349]: the `end|else` regexp is not quite right as the `\b` is only on the `else`. Needs parens like the other.

File: frontend/module/editor.js [hunk 2]
```diff
@@ -343,25 +346,68 @@ const preserveIndent = () => (textareaProps, event) => {
...
+  if (matches[0] === "") {
+    return
+  }
+  if (selectionStart !== selectionEnd && matches[4] !== undefined) {
+    // fixing current line indent for 'end' and 'else' with selection is too hard.
     return
   }
 
+  console.log("start", selectionStart, "end", selectionEnd)
+  console.log("matches", matches)
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply frontend/module/editor.js:359]: absolutely.
not.
thanks for pointing it out.

File: frontend/module/editor.js (hunk 1 above)
Comment (juliaogris) [reply frontend/module/editor.js:349]: whoops. I <3 regexp

Comment (juliaogris) [reply frontend/module/editor.js:349]: thank you.

File: frontend/module/editor.js [hunk 3]
```diff
@@ -343,25 +346,68 @@ const preserveIndent = () => (textareaProps, event) => {
...
     value: newValue,
-    selectionStart: selectionStart + inserted.length,
-    selectionEnd: selectionStart + inserted.length,
+    selectionStart: newIdx,
+    selectionEnd: newIdx,
+  }
+}
+
+function shouldDedent(indent, lines, currentLineNumber) {
+  if (indent.length < 4 || currentLineNumber === 0) {
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply frontend/module/editor.js:405]: ok as discussed, I've updated to `indent.lenth == 0` and modified the use to:
```
   if (shouldDedent(indent, lines, currentLineNumber)) {
      const dedent = Math.min(4, indent.length)
      const beforeCurrentLine = before.substring(0, before.length - currentLine.length)
      const dedentedCurrentLine = currentLine.substring(dedent)
      before = beforeCurrentLine + dedentedCurrentLine
      // Decrease indent
      indent = indent.substring(dedent)
    }
```

File: frontend/module/editor.js [hunk 4]
```diff
@@ -343,25 +346,68 @@ const preserveIndent = () => (textareaProps, event) => {
...
-
-  const inserted = newLine + indent
-
-  const newValue = value.substring(0, selectionStart) + inserted + value.substring(selectionEnd)
+  const indent = matches[1] || ""
+  const before = value.substring(0, selectionStart)
+  const after = value.substring(selectionEnd)
+  let newValue
+
+  let inserted
```
[the remark this answers is not in the document]
Comment (juliaogris) [reply frontend/module/editor.js:366]: done.

File: frontend/module/editor.js (hunk 1 above)
Comment (juliaogris) [reply frontend/module/editor.js:349]: done.

**Review by camh- — APPROVED** [review]
:green_apple:

**Pull request merged**